src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -670,13 +671,70 @@ internal override void Load(XmlElement root)
                         {
                             var providerConfig = new ProviderCategoryConfiguration();
                             providerConfig.Load(child);
-                            ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+
+                            if (ProviderConfigurations.ContainsKey(providerConfig.Name))
+                            {
+                                var existingProviderConfig = ProviderConfigurations[providerConfig.Name];
+                                existingProviderConfig.Merge(providerConfig);
+                            }
+                            else
+                            {
+                                ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+                            }
                         }
                         break;
                 }
             }
         }
 
+        /// <summary>
+        /// Registers given type of <typeparamref name="T"/> as bootstrap provider
+        /// </summary>
+        /// <typeparam name="T">Non-abstract type which implements <see cref="IBootstrapProvider"/> interface</typeparam>
+        /// <param name="properties">Properties that will be passed to bootstrap provider upon initialization</param>
+        public void RegisterBootstrapProvider<T>(IDictionary<string, string> properties = null)
```

Comment:
Could be. Thus, it won't add anything. And it will only add additional noise to client's code (ie .ToReadOnly()).  ImmutableDictionary is also an option. Another option is IEnumerable<KeyValuePair<string, string>>